sway-core/src/semantic_analysis/namespace/namespace.rs

```diff
@@ -220,14 +225,62 @@ impl Namespace {
                 errors
             );
             if &method.name == method_name {
-                // if we find the method that we are looking for, we also need
-                // to retrieve the impl definitions for the return type so that
-                // the user can string together method calls
-                self.insert_trait_implementation_for_type(engines, method.return_type);
-                return ok(decl_id, warnings, errors);
+                matching_method_decl_ids.push(decl_id);
             }
         }
 
+        let matching_method_decl_id = match matching_method_decl_ids.len().cmp(&1) {
+            Ordering::Equal => Some(matching_method_decl_ids[0].clone()),
+            Ordering::Greater => {
+                // Case where multiple methods exist with the same name
+                // This is the case of https://github.com/FuelLabs/sway/issues/3633
+                // where multiple generic trait impls use the same method name but with different parameter types
+                let mut matching_method_decl_ids2: Vec<DeclarationId> = vec![];
```

Comment:
This doesn't need to be a list---could you refactor this to an `Option<DeclarationId>`?

This can be refactored because the only functions that this variable is used for is to check that is (1) has been set and (2) get the first element from the list. What you could do instead is to use an `Option<DeclarationId>` called `maybe_method_decl_id`. When `maybe_method_decl_id` is `None`, then we know that we haven't found an end condition yet, but when we find that end condition (i.e. the large conditional in the `if` expression below) we can set this to `Some(decl_id)`, when that happens we can break out of the loop and avoid wasted compute.

Reply:
Good point. Applied the changes suggested.